src/libponyrt/actor/actor.c

```diff
@@ -432,69 +438,124 @@ bool ponyint_actor_run(pony_ctx_t* ctx, pony_actor_t* actor, bool polling)
-    // action. Therefore if `noblock` is on, we should garbage collect the
-    // actor.
-    //
-    // When the cycle detector is running, "deleting locally" is still far
-    // more efficient and will free up memory sooner than letting the cycle
-    // detector do it. However, that is only safe to delete locally if the
-    // cycle detector isn't holding a reference to the actor. If we have never
-    // sent a reference to the cycle detector (FLAG_CD_CONTACTED), then this
-    // is perfectly safe to do.
-    if (actor_noblock || !has_flag(actor, FLAG_CD_CONTACTED))
+
+    // check if the actors queue is empty or not
+    if(ponyint_messageq_isempty(&actor->q))
     {
-      // "Locally delete" the actor.
-      ponyint_actor_setpendingdestroy(actor);
-      ponyint_actor_final(ctx, actor);
-      ponyint_actor_destroy(actor);
-    } else {
-      // Tell cycle detector that this actor is a zombie and will not get
-      // any more messages/work and can be reaped.
-      // Mark the actor as FLAG_BLOCKED_SENT and send a BLOCKED message
-      // to speed up reaping otherwise waiting for the cycle detector
-      // to get around to asking if we're blocked could result in unnecessary
-      // memory growth.
-      if(!has_flag(actor, FLAG_BLOCKED_SENT))
+      // The actors queue is empty which means this actor is a zombie
+      // and can be reaped.
+      // If the cycle detector is disabled (or it doesn't know about this actor),
+      // this actors queue is guaranteed to remain empty after this point as no
+      // other actor has references to it.
+      // If the cycle detector is enabled and knows about this actor,
+      // then it can possibly add messages onto the actors queue making
+      // it not safe to be destroyed.
+      //
+      // As a result, how reaping can occur depends on whether the cycle
+      // detector is running whether it is holding a reference to the actor
+      // or not.
+      //
+      // When 'actor_noblock` is true, the cycle detector isn't running.
+      // this means actors won't be garbage collected unless we take special
+      // action. Therefore if `noblock` is on, we should garbage collect the
+      // actor.
+      //
+      // When the cycle detector is running, "deleting locally" is still far
+      // more efficient and will free up memory sooner than letting the cycle
+      // detector do it. However, that is only safe to delete locally if the
+      // cycle detector isn't holding a reference to the actor. If we have never
+      // sent a reference to the cycle detector (FLAG_CD_CONTACTED), then this
+      // is perfectly safe to do.
+      if (actor_noblock || !has_flag(actor, FLAG_CD_CONTACTED))
       {
-        // We're blocked, send block message.
+        // mark the queue as empty or else destroy will hang
+        bool empty = ponyint_messageq_markempty(&actor->q);
+
+        // make sure the queue is actually empty as expected
```

Comment:
I think we should restate the assumptions you and I have privately spoken about here.

Those points might not be clear to others in the future.

- If no block and rc of 0 then nothing could have sent a message since we did the empty check.
- If not no block, but rc of 0 and cd has never been contacted, nothing could have sent a message since we did the empty check

someone who isn't